### DISCUSSION TITLE
Cryptographic signing of plans

### DISCUSSION THREAD

**Discussion by sandstrom** [opentofu/opentofu#1245] [Ideas]
A workflow like this would be great:

1. A PR is opened to perform some infrastructure change.
2. The plan is signed by a local key (perhaps even on a Yubikey or similar) on an engineers computer. This may be the PR author, or a reviewer. Perhaps the system could optionally piggy-back on Git commit signatures(?).
3. After the PR is accepted, the plan is passed along (exactly how can vary) to a privileged machine with sufficient access rights to apply the plan. This machine will verify that the plan is signed with a trusted signature. Thereafter it's applied.

The benefit with this approach is that third-party services such as Spacelift, Env0 and the others can provide a great UI and workflow, but doesn't have to be trusted. If they are hacked, no access is lost. 

It's similar to other efforts in secure development, where the CI/CD flow shouldn't necessarily be trusted, and instead cryptographic guarantees should be used (https://security.googleblog.com/2021/06/introducing-slsa-end-to-end-framework.html).

I know this can also be hacked together outside of OpenTofu. But it's something that I think would be more useful if it was standardised and implemented in the CLI tooling.

### Implementation details

- Should it be possible to require more than one signature?
- Should the signature be packaged inside the JSON plan output, or kept separate?

Tracked in https://github.com/opentofu/opentofu/issues/1246

**Comment by ghost** [comment]
Hi @sandstrom could you please open this as an issue?

**Reply by sandstrom** [reply]
Added here: https://github.com/opentofu/opentofu/issues/1246
